### PR TITLE
Temperature-dependent elastic constants in tensor_mechanics

### LINKED ISSUE #8265: Temperature-dependent elastic constants in tensor_mechanics
### Description of the enhancement or error report
We need the ability to prescribe Young's modulus and Poisson's ratio using temperature-dependent functions in tensor_mechanics. This capability currently exists only in solid_mechanics.

### Rationale for the enhancement or information for reproducing the error
This is an important capability needed to convert Grizzly reactor pressure vessel models over to tensor_mechanics.

### Identified impact
Add a new elasticity tensor material model for this purpose.

### PR DESCRIPTION
**Pull request by bwspenc**
closes #8265

### CODE REVIEW

File: modules/tensor_mechanics/src/materials/ComputeTemperatureDependentIsotropicElasticityTensor.C [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+void
+ComputeTemperatureDependentIsotropicElasticityTensor::computeQpElasticityTensor()
+{
+  const Point p;
+  const Real youngs_modulus = _youngs_modulus_function.value(_temperature[_qp], p);
+  const Real poissons_ratio = _poissons_ratio_function.value(_temperature[_qp], p);
+
+  // Build the vector to pass into RankFourTensor to create the elasticity tensor
+  std::vector<Real> isotropic_elastic_constants(2);
```
Comment (dschwen) [inline modules/tensor_mechanics/src/materials/ComputeTemperatureDependentIsotropicElasticityTensor.C]: Not a super big deal, but local vector variables introduce quite a bit of memory churn and affect performance. I suggest making that vector a member.

Comment (dschwen) [inline modules/tensor_mechanics/src/materials/ComputeTemperatureDependentIsotropicElasticityTensor.C]: same with the point

File: modules/tensor_mechanics/src/materials/ComputeTemperatureDependentIsotropicElasticityTensor.C [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+#include "Function.h"
+
+template<>
+InputParameters validParams<ComputeTemperatureDependentIsotropicElasticityTensor>()
+{
+  InputParameters params = validParams<ComputeElasticityTensorBase>();
+  params.addClassDescription("Compute an isotropic elasticity tensor for elastic constants that change as a function of temperature");
+  params.addRequiredCoupledVar("temperature", "Coupled temperature");
+  params.addRequiredParam<FunctionName>("youngs_modulus_function","Function defining Young's modulus as a function of temperature");
+  params.addRequiredParam<FunctionName>("poissons_ratio_function","Function defining Poisson's ratio as a function of temperature");
```
Comment (dschwen) [inline modules/tensor_mechanics/src/materials/ComputeTemperatureDependentIsotropicElasticityTensor.C]: Worlds collide here. In tensor mechanics (or phase field) we do not use functions in that way anywhere. I would have implemented this as a material property (and allowed for arbitrary variable dependencies). There is nothing that is specific to "Temperature" in this class at all.

The way this class is implemented users might be tempted to abuse it (in absence of a `ComputeVariableIsotropicElasticityTensor` class). The problem is that this class does not provide derivatives of the elasticity tensor, which has the potential to break stuff.

File: modules/tensor_mechanics/src/materials/ComputeTemperatureDependentIsotropicElasticityTensor.C (hunk 1 above)
Comment (friedmud) [inline modules/tensor_mechanics/src/materials/ComputeTemperatureDependentIsotropicElasticityTensor.C]: `Point` objects are statically sized and stack-allocated.  They really shouldn't create memory churn.  See:

https://github.com/libMesh/libmesh/blob/master/include/numerics/type_vector.h#L386

I agree about the `std::vector` though.

Comment (permcody) [inline modules/tensor_mechanics/src/materials/ComputeTemperatureDependentIsotropicElasticityTensor.C]: If you want to leave it local, this is a textbook case for `std::array`. Unlike `std::vector`, arrays are statically sized and stack-allocated.

**Comment by bwspenc** [comment]
I'm closing this in favor of the new version I'm proposing in PR #8613.

**Pull request closed without merging**